File: python/61_Rotate_List.py
# coding=utf-8
# 61
import unittest
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution:
    def solve(self, head, k):
        return self.rotateRight(head, k)
    # @param {ListNode} head
    # @param {integer} k
    # @return {ListNode}
    def rotateRight(self, head, k):
        if not head or k == 0:
            return head
        runner=cur=head
        length=0
        while runner:
            runner=runner.__next__
            length+=1
        runner=head
        k=k%length
        if k==0:
            return head
        while k>0:
            runner=runner.__next__
            k-=1
        while runner and runner.__next__:
            cur=cur.__next__
            runner=runner.__next__
        res=cur.__next__
        runner.next=head
        cur.next=None
        return res
        
class Test(unittest.TestCase):
    #override
    def setUp(self):
        self.s=Solution()
        
    def tearDown(self):
        print("TestComplete")
        
    def genTestCase(self):
        case=[]
        num=0
        head=ListNode(num)
        for i in range(num-1,-1,-1):
            tmp=ListNode(i)
            tmp.next=head
            head=tmp
        case.append(head)
        while head:
            print(head.val, end=' ')
            head=head.__next__
        print() 
        self.cases=case
        
    def test(self):
        self.genTestCase()
        for case in self.cases:
            print("testcase: ", case)
            print("result: ",self.s.solve(case,1))
            # if assert is needed
            
unittest.main()

File: Java/61_Rotate_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    // O(2n)
    public ListNode rotateRight(ListNode head, int k) {
        if(head==null||head.next==null||k==0)return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode runner = head, slow = dummy;
        int len=1;
        // 找到链表尾部，顺便计算链表长度
        while(runner.next!=null){
            runner=runner.next;
            len++;
        }
        k=len-k%len;// 计算前面要留出来的长度
        // 找到要移动的第一个节点的前一个节点
        while(k>0){
            slow = slow.next;
            k--;
        }
        runner.next = dummy.next;
        dummy.next = slow.next;
        slow.next = null;
        return dummy.next;
    }
}